### PR TITLE
Adaptive kernel Kalman filter (AKKF)

### PR DESCRIPTION
**Pull request by jswright-dstl**
This PR adds the Adaptive kernel Kalman filter (AKKF) [1]. The implementation includes the following:
- Kernel classes
  - `QuadraticKernel` 
  - `QuarticKernel`
  - `GaussianKernel`
- `AdaptiveKernelKalmanPredictor`
- `AdaptiveKernelKalmanUpdater`
- `KernelParticleState`
- AKKF Tutorial

The above implementations are described in [2]. 

[1] M. Sun, M. E. Davies, I. Proudler and J. R. Hopgood, "Adaptive Kernel Kalman Filter," 2021 Sensor Signal Processing for Defence Conference (SSPD), Edinburgh, United Kingdom, 2021, pp. 1-5, doi: 10.1109/SSPD51364.2021.9541455.
[2] J. S. Wright, J. R. Hopgood, M. E. Davies, I. K. Proudler and M. Sun, "Implementation of Adaptive Kernel Kalman Filter in Stone Soup," 2023 Sensor Signal Processing for Defence Conference (SSPD), Edinburgh, United Kingdom, 2023, pp. 1-5, doi: 10.1109/SSPD57945.2023.10256739.

### CODE REVIEW

File: stonesoup/updater/kernel.py [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+    """Adaptive kernel Kalman updater class
+    """
+    kernel: Kernel = Property(
+        default=None,
+        doc="Kernel")
+    lambda_updater: float = Property(
+        default=1e-3,
+        doc="Used to incorporate prior knowledge of the distribution. If the "
+            "true distribution is Gaussian, the value of 2 is optimal. "
+            "Default is 2")
```
Comment (sdhiscocks) [inline stonesoup/updater/kernel.py:24]: Default doesn't match doc.

File: stonesoup/updater/kernel.py [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+from ..types.update import Update
+from ..base import Property
+
+
+class AdaptiveKernelKalmanUpdater(Updater):
+    """Adaptive kernel Kalman updater class
+    """
+    kernel: Kernel = Property(
+        default=None,
+        doc="Kernel")
```
Comment (sdhiscocks) [inline stonesoup/updater/kernel.py:19]: Document default when `None`.

File: stonesoup/types/state.py [hunk 3]
```diff
@@ -917,6 +917,44 @@ def __getitem__(self, item):
...
+        """The number of dimensions represented by the state."""
+        return self.state_vector.shape[0]
+
+    @property
+    def mean(self):
+        return self.state_vector @ self.weight
+
+    @property
+    def covar(self):
+        return self.state_vector @ self.kernel_covar @ self.state_vector.T
```
Comment (sdhiscocks) [inline stonesoup/types/state.py:952]:
```suggestion
    @clearable_cached_property('state_vector', 'weight')
    def mean(self):
        return self.state_vector @ self.weight

    @clearable_cached_property('state_vector', 'kernel_covar')
    def covar(self):
        return self.state_vector @ self.kernel_covar @ self.state_vector.T
```

Comment (sdhiscocks) [inline stonesoup/types/state.py:952]: No weight for covar?

File: docs/tutorials/filters/AKKF.py [hunk 4]
```diff
@@ -0,0 +1,381 @@
...
+
+plotter = Plotter()
+plotter.plot_ground_truths(truth, [0, 2], linewidth=3.0, color='black')
+plotter.plot_tracks(track, [0, 2], track_label='AKKF - quadratic', color='royalblue')
+plotter.fig
+
+# %%
+
+plotter.plot_tracks(track, [0, 2], particle=True)
+plotter.fig
```
Comment (sdhiscocks) [inline docs/tutorials/filters/AKKF.py:381]: Adding the comment below should use last image for gallery thumbnail, which I think is better.

```suggestion
# %%

# sphinx_gallery_thumbnail_number = -1

plotter.plot_tracks(track, [0, 2], track_label='AKKF - quadratic', color='royalblue', particle=True)
plotter.fig
```

File: stonesoup/kernel.py [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+
+class Kernel(Base):
+    """Kernel base type
+
+    A Kernel provides a means to translate state space or measurement space into kernel space.
+    """
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
```
[the remark this answers is not in the document]
Comment (jswright-dstl) [reply stonesoup/kernel.py]: Removed.

File: stonesoup/kernel.py [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+            state_vector1 = state1
+        if state2 is None:
+            state_vector2 = state_vector1
+        else:
+            if isinstance(state2, State):
+                state_vector2 = state2.state_vector
+            else:
+                state_vector2 = state2
+        return (state_vector1.T @ state_vector2 /
+                self.ialpha + self.c) ** 4
```
[the remark this answers is not in the document]
Comment (jswright-dstl) [reply stonesoup/kernel.py]: Changed.

File: stonesoup/kernel.py [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+            state_vector1 = state1.state_vector
+        else:
+            state_vector1 = state1
+        if state2 is None:
+            state_vector2 = state_vector1
+        else:
+            if isinstance(state2, State):
+                state_vector2 = state2.state_vector
+            else:
+                state_vector2 = state2
```
[the remark this answers is not in the document]
Comment (jswright-dstl) [reply stonesoup/kernel.py]: Moved to private method in base Kernel class.

File: stonesoup/kernel.py [hunk 8]
```diff
@@ -0,0 +1,183 @@
...
+        diff_tilde_x = (state_vector1.T[:, :, None] - state_vector2.T[:, None, :]) ** 2
+        diff_tilde_x_sum = np.sum(diff_tilde_x, axis=0)
+
+        k_tilde_x = np.exp(-diff_tilde_x_sum / (2 * self.variance)) / (
+            np.sqrt(2 * np.pi * self.variance))
```
[the remark this answers is not in the document]
Comment (jswright-dstl) [reply stonesoup/kernel.py]: Changed.

File: stonesoup/predictor/kernel.py [hunk 9]
```diff
@@ -0,0 +1,83 @@
+import copy
+import numpy as np
```
[the remark this answers is not in the document]
Comment (jswright-dstl) [reply stonesoup/predictor/kernel.py]: Added.

File: stonesoup/predictor/kernel.py [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+        doc=r":math:`\lambda_{\tilde{K}}`. Regularisation parameter used to stabilise the inverse "
+            r"Gram matrix. Range is :math:`\left[10^{-4}, 10^{-2}\right]`")
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+        self._time_interval = None
+
+        if self.kernel is None:
+            self.kernel = QuadraticKernel()
```
[the remark this answers is not in the document]
Comment (jswright-dstl) [reply stonesoup/predictor/kernel.py]: Removed.

File: stonesoup/predictor/kernel.py [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+        -------
+        : :class:`~.KernelParticleStatePrediction`
+            The predicted state :math:`\mathbf{x}_{k|k-1}` and the predicted
+            state covariance :math:`P_{k|k-1}`
+        """
+        if proposal is None:
+            if isinstance(prior, KernelParticleStateUpdate):
+                proposal = State(state_vector=prior.proposal)
+            else:
+                proposal = copy.deepcopy(prior)
```
[the remark this answers is not in the document]
Comment (jswright-dstl) [reply stonesoup/predictor/kernel.py]: Changed to copy.

File: stonesoup/predictor/kernel.py [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+        k_tilde_tilde = self.kernel(proposal)
+        k_tilde_nontilde = self.kernel(proposal, prior)
+
+        kernel_t = np.linalg.pinv(
+            k_tilde_tilde + self.lambda_predictor * np.identity(len(prior))) @ k_tilde_nontilde
+        prediction_weights = kernel_t @ prior.weight
+        v = (np.linalg.pinv(k_tilde_tilde + self.lambda_predictor * np.identity(
+            len(prior))) @ k_tilde_tilde - np.identity(len(prior))) @ \
+            (np.linalg.pinv(k_tilde_tilde + self.lambda_predictor * np.identity(
+                len(prior))) @ k_tilde_tilde - np.identity(len(prior))).T / len(prior)
```
[the remark this answers is not in the document]
Comment (jswright-dstl) [reply stonesoup/predictor/kernel.py]: Happy with the tidy up of the equations. Changed.

File: stonesoup/updater/kernel.py [hunk 13]
```diff
@@ -0,0 +1,108 @@
...
+    def predict_measurement(self, state_prediction, measurement_model=None,
+                            **kwargs):
+
+        if measurement_model is None:
+            measurement_model = self.measurement_model
+
+        new_state_vector = measurement_model.function(state_prediction, **kwargs)
+        return MeasurementPrediction.from_state(
+            state_prediction,
+            state_vector=new_state_vector)
```
[the remark this answers is not in the document]
Comment (jswright-dstl) [reply stonesoup/updater/kernel.py:42]: Added `measurement_noise`.

File: stonesoup/updater/kernel.py [hunk 14]
```diff
@@ -0,0 +1,108 @@
...
+        g_y = self.kernel(hypothesis.measurement_prediction, hypothesis.measurement)
+
+        Q_AKKF = predicted_state.kernel_covar @ np.linalg.pinv(
+            G_yy @ predicted_state.kernel_covar + self.lambda_updater * np.identity(
+                len(predicted_state)))
+
+        updated_weights = np.atleast_2d(predicted_state.weight).T + Q_AKKF @ (
+                    g_y - np.atleast_2d(G_yy @ predicted_state.weight).T)
+        updated_covariance = \
+            predicted_state.kernel_covar - Q_AKKF @ G_yy @ predicted_state.kernel_covar
```
[the remark this answers is not in the document]
Comment (jswright-dstl) [reply stonesoup/updater/kernel.py]: Changed.

File: stonesoup/predictor/kernel.py [hunk 15]
```diff
@@ -0,0 +1,83 @@
...
+from ..types.state import State
+from ..types.update import KernelParticleStateUpdate
+
+
+class AdaptiveKernelKalmanPredictor(KalmanPredictor):
+    """AdaptiveKernelKalmanFilter class
+    """
+    kernel: Kernel = Property(
+        default=None,
+        doc="Kernel")
```
[the remark this answers is not in the document]
Comment (jswright-dstl) [reply stonesoup/predictor/kernel.py]: Added.

File: stonesoup/predictor/kernel.py [hunk 16]
```diff
@@ -0,0 +1,83 @@
...
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+        self._time_interval = None
+
+        if self.kernel is None:
+            self.kernel = QuadraticKernel()
+
+    @predict_lru_cache()
+    def predict(self, prior, proposal=None, timestamp=None, **kwargs):
```
[the remark this answers is not in the document]
Comment (jswright-dstl) [reply stonesoup/predictor/kernel.py]: Changed.

File: stonesoup/updater/kernel.py (hunk 1 above)
Comment (jswright-dstl) [reply stonesoup/updater/kernel.py]: Changed doc.

File: stonesoup/updater/kernel.py (hunk 2 above)
Comment (jswright-dstl) [reply stonesoup/updater/kernel.py]: Added.

File: stonesoup/kernel.py [hunk 17]
```diff
@@ -0,0 +1,183 @@
...
+            state_vector1 = state1
+        if state2 is None:
+            state_vector2 = state_vector1
+        else:
+            if isinstance(state2, State):
+                state_vector2 = state2.state_vector
+            else:
+                state_vector2 = state2
+        return (state_vector1.T @ state_vector2 /
+                self.ialpha + self.c) ** 2
```
[the remark this answers is not in the document]
Comment (jswright-dstl) [reply stonesoup/kernel.py]: Changed.

**Review by sdhiscocks — CHANGES_REQUESTED** [review]
Minor doc suggestions, and change to which variable being used to reset caching.

Also found issue with weights of state being inconsistent, so proposed below to fix that.

```diff
diff --git a/stonesoup/predictor/kernel.py b/stonesoup/predictor/kernel.py
index e329809d..cdcd1361 100644
--- a/stonesoup/predictor/kernel.py
+++ b/stonesoup/predictor/kernel.py
@@ -76,7 +76,7 @@ class AdaptiveKernelKalmanPredictor(KalmanPredictor):
         prediction_covariance = kernel_t @ prior.kernel_covar @ kernel_t.T + v
         return Prediction.from_state(prior,
                                      state_vector=new_state_vector,
-                                     weight=np.squeeze(prediction_weights),
+                                     weight=prediction_weights,
                                      kernel_covar=prediction_covariance,
                                      timestamp=timestamp,
                                      transition_model=self.transition_model)
diff --git a/stonesoup/types/state.py b/stonesoup/types/state.py
index f1965f1b..ea116e7f 100644
--- a/stonesoup/types/state.py
+++ b/stonesoup/types/state.py
@@ -945,7 +945,7 @@ class KernelParticleState(State):
 
     @clearable_cached_property('state_vector', 'weight')
     def mean(self):
-        return self.state_vector @ self.weight
+        return self.state_vector @ self.weight[:, np.newaxis]
 
     @clearable_cached_property('state_vector', 'weight')
     def covar(self):
diff --git a/stonesoup/types/tests/test_state.py b/stonesoup/types/tests/test_state.py
index 3fd80f69..eba966ce 100644
--- a/stonesoup/types/tests/test_state.py
+++ b/stonesoup/types/tests/test_state.py
@@ -863,5 +863,5 @@ def test_kernel_particle_state():
     assert np.array_equal(prior.kernel_covar, prior_w_kernel_covar.kernel_covar)
     assert number_particles == len(prior)
     assert 4 == prior.ndim
-    assert np.array_equal(state_vector @ weights, prior.mean)
+    assert np.array_equal(state_vector @ weights[:, np.newaxis], prior.mean)
     assert np.array_equal(state_vector @ np.diag(weights) @ state_vector.T, prior.covar)
diff --git a/stonesoup/updater/kernel.py b/stonesoup/updater/kernel.py
index 9860675a..e9998377 100644
--- a/stonesoup/updater/kernel.py
+++ b/stonesoup/updater/kernel.py
@@ -86,14 +86,13 @@ class AdaptiveKernelKalmanUpdater(Updater):
             predicted_state.kernel_covar \
             @ np.linalg.pinv(G_yy @ predicted_state.kernel_covar
                              + self.lambda_updater * np.identity(len(predicted_state)))
-        updated_weights = \
-            np.atleast_2d(predicted_state.weight).T \
-            + Q_AKKF @ (g_y - np.atleast_2d(G_yy @ predicted_state.weight).T)
+        weights = predicted_state.weight[:, np.newaxis]
+        updated_weights = (weights + Q_AKKF@(g_y - G_yy@weights)).ravel()
         updated_covariance = \
             predicted_state.kernel_covar - Q_AKKF @ G_yy @ predicted_state.kernel_covar
 
         # Proposal Calculation
-        pred_mean = predicted_state.state_vector @ np.squeeze(updated_weights)
+        pred_mean = predicted_state.state_vector @ updated_weights
         pred_covar = np.diag(np.diag(
             predicted_state.state_vector @ updated_covariance @ predicted_state.state_vector.T))
 
diff --git a/stonesoup/updater/tests/test_kernel.py b/stonesoup/updater/tests/test_kernel.py
index ceea0c64..06c671f7 100644
--- a/stonesoup/updater/tests/test_kernel.py
+++ b/stonesoup/updater/tests/test_kernel.py
@@ -108,7 +108,7 @@ def test_kernel_updater(kernel, measurement_model, c, ialpha):
     assert update.hypothesis.measurement.timestamp == gt_state.timestamp
     assert np.allclose(update.state_vector, prediction.state_vector)
     assert np.allclose(update.proposal, StateVectors(new_state_vector.T), atol=1e0)
-    assert np.allclose(update.weight, updated_weights)
+    assert np.allclose(update.weight, updated_weights.ravel())
     assert np.allclose(update.kernel_covar, updated_covariance)
 
 
```

**Review by sdhiscocks — APPROVED** [review]

File: stonesoup/predictor/kernel.py [hunk 18]
```diff
@@ -0,0 +1,82 @@
...
+from .kalman import KalmanPredictor
+from ..base import Property
+from ..kernel import Kernel, QuadraticKernel
+from ..types.prediction import Prediction
+from ..types.state import State
+from ..types.update import KernelParticleStateUpdate
+
+
+class AdaptiveKernelKalmanPredictor(KalmanPredictor):
+    """AdaptiveKernelKalmanFilter class
```
Comment (hpritchett-dstl) [inline stonesoup/predictor/kernel.py:15]: would be useful to expand on this docstring to be consistent with documentation for other predictors.

File: stonesoup/updater/kernel.py [hunk 19]
```diff
@@ -0,0 +1,109 @@
...
+from . import Updater
+from ..kernel import QuadraticKernel, Kernel
+from ..types.array import StateVectors
+from ..types.prediction import MeasurementPrediction
+from ..types.update import Update
+from ..base import Property
+
+
+class AdaptiveKernelKalmanUpdater(Updater):
+    """Adaptive kernel Kalman updater class
```
Comment (hpritchett-dstl) [inline stonesoup/updater/kernel.py:15]: would be useful to expand on this docstring to be consistent with documentation for other updaters.

**Review by hpritchett-dstl — APPROVED** [review]

**Pull request merged**